### PR TITLE
Added required field checking into the encoder

### PR DESCRIPTION
**Pull request by haberman**

### CODE REVIEW

File: upb/msg_internal.h [hunk 1]
```diff
@@ -198,6 +198,18 @@ typedef struct {
   int ext_count;
 } upb_msglayout_file;
 
+// Computes a bitmask in which the |n| lowest bits are set, except that we
```
Comment (fowles) [inline upb/msg_internal.h:201]: the comment says `n` but that is a derived quantity.  You should make it say `l->required_count`

File: upb/encode.c [hunk 2]
```diff
@@ -511,6 +511,15 @@ static void encode_message(upb_encstate *e, const upb_msg *msg,
                            const upb_msglayout *m, size_t *size) {
   size_t pre_len = e->limit - e->ptr;
 
+  if (e->options & UPB_ENCODE_CHECKREQUIRED && m->required_count) {
```
[the remark this answers is not in the document]
Comment (haberman) [reply upb/encode.c:514]: Done.

File: upb/msg_internal.h [hunk 3]
```diff
@@ -198,6 +198,18 @@ typedef struct {
...
+// skip the lowest bit (because upb never uses hasbit 0).
+//
+// Sample output:
+//    requiredmask(1) => 0b10 (0x2)
+//    requiredmask(5) => 0b111110 (0x3e)
+UPB_INLINE uint64_t upb_msglayout_requiredmask(const upb_msglayout *l) {
+  int n = l->required_count;
+  assert(0 < n && n < 63);
+  return ((1 << n) - 1) << 1;
```
[the remark this answers is not in the document]
Comment (haberman) [reply upb/msg_internal.h:210]: Done.

File: upb/msg_internal.h (hunk 1 above)
Comment (haberman) [reply upb/msg_internal.h:201]: Done.

**Comment by haberman** [comment]
PTAL.

File: BUILD [hunk 4]
```diff
@@ -230,10 +230,27 @@ cc_test(
...
+    name = "gen_max_required_test_proto",
+    outs = ["upb/max_required_test.proto"],
+    cmd = """
+          echo "syntax = \\\"proto2\\\";" >> $@
+          echo "package upb_test;" >> $@
+          echo "message TestMaxRequiredFields {" >> $@
+          for i in {1..63}; do
+            echo "  required int32 required_int32_$$i = $$i;" >> $@
+          done
+          echo "}" >> $@
```
Comment (fowles) [inline BUILD:250]: a genrule feels fairly heavy weight for this.  I would simply gen it out once and commit the proto file.

File: upb/msg_internal.h [hunk 5]
```diff
@@ -198,16 +198,16 @@ typedef struct {
...
 //
 // Sample output:
 //    requiredmask(1) => 0b10 (0x2)
 //    requiredmask(5) => 0b111110 (0x3e)
 UPB_INLINE uint64_t upb_msglayout_requiredmask(const upb_msglayout *l) {
   int n = l->required_count;
-  assert(0 < n && n < 63);
-  return ((1 << n) - 1) << 1;
+  assert(0 < n && n <= 63);
+  return (((uint64_t)1 << n) - 1) << 1;
```
Comment (fowles) [inline upb/msg_internal.h:210]: oh god, now I have to remember the precedence of cast vs `<<`...  Can We get another set of parens?  The expression is getting a bit ridiculous I know....  Alternately you could break it into multiple statements:

```
uint64_t r = 1;
r <<= n;
r -= 1;
r << = 1;
return r;
```

I don't love it so feel free to go either way

**Review by fowles — APPROVED** [review]

File: upb/msg_internal.h (hunk 5 above)
Comment (haberman) [reply upb/msg_internal.h]: I changed it to `1ULL`.

File: BUILD (hunk 4 above)
Comment (haberman) [reply BUILD]: Done.

**Comment by haberman** [comment]
PTAL.

**Comment by haberman** [comment]
Ah nvm, I just saw that you approved previously, thanks!

**Pull request merged**